dsc_lib/src/discovery/command_discovery.rs

```diff
@@ -71,83 +51,136 @@ impl ResourceDiscovery for CommandDiscovery {
+                            {
+                                resources.insert(resource.type_name.clone(), resource);
+                            }
+                            else if remaining_required_resource_types.contains(&resource.type_name)
+                            {
+                                remaining_required_resource_types.retain(|x| *x != resource.type_name);
+                                debug!("Found {} in {}", &resource.type_name, path.display());
+                                resources.insert(resource.type_name.clone(), resource);
+                                if remaining_required_resource_types.is_empty()
+                                {
+                                    return Ok(resources);
                                 }
                             }
-                            self.resources.insert(resource.type_name.clone(), resource.clone());
                         }
                     }
                 }
             }
         }
 
+        debug!("Found {} matching non-provider resources", resources.len() - provider_resources.len());
+
         // now go through the provider resources and add them to the list of resources
-        for provider in &self.provider_resources {
-            let provider_resource = self.resources.get(provider).unwrap();
+        for provider in provider_resources {
+            debug!("Enumerating resources for provider {}", provider);
+            let provider_resource = resources.get(&provider).unwrap();
             let provider_type_name = provider_resource.type_name.clone();
-            let provider_path = provider_resource.path.clone();
             let manifest = import_manifest(provider_resource.manifest.clone().unwrap())?;
+            let mut provider_resources_count = 0;
             // invoke the list command
             let list_command = manifest.provider.unwrap().list;
             let (exit_code, stdout, stderr) = match invoke_command(&list_command.executable, list_command.args, None, Some(&provider_resource.directory), None)
             {
                 Ok((exit_code, stdout, stderr)) => (exit_code, stdout, stderr),
                 Err(e) => {
-                    self.discovery_messages.push(StreamMessage::new_error(
-                        format!("Could not start {}: {}", list_command.executable, e),
-                        Some(provider_type_name.clone()),
-                        Some(provider_path.clone())));
-
+                    // In case of "resource list" operation - print failure from provider as warning
+                    // In case of other resource/config operations:
+                    // print failure from provider as error because this provider was specifically requested by current resource/config operation
+                    if return_all_resources { warn!("Could not start {}: {}", list_command.executable, e); } else { error!("Could not start {}: {}", list_command.executable, e); }
```

Comment:
follow the style convention and put on multiple lines

Reply:
Updated.